File: fourSum.go
package main

func FourNumberSum(array []int, target int) [][]int {
	quadruplets := [][]int{}

	paritalSums := make(map[int][][]int)

	for i := 1; i < len(array)-2; i++ {
		for k := 0; k < i; k++ {
			currentSum := array[i] + array[k]
			paritalSums[currentSum] = append(paritalSums[currentSum], []int{array[i], array[k]})
		}
		for j := i + 2; j < len(array); j++ {
			currentSum := array[i+1] + array[j]
			diff := target - currentSum
			if pairs, found := paritalSums[diff]; found {
				for _, pair := range pairs {
					quadruplets = append(quadruplets, append(pair, array[i+1], array[j]))
				}
			}
		}
	}

	return quadruplets
}

func FourNumberSumV3(array []int, target int) [][]int {
	quadruplets := [][]int{}

	paritalSums := map[int][][]int{}

	for i := 0; i < len(array)-1; i++ {
		for j := i + 1; j < len(array); j++ {
			difference := target - (array[i] + array[j])
			values, found := paritalSums[difference]
			if found {
				for _, v := range values {
					quadruplets = append(quadruplets, []int{array[i], array[j], v[0], v[1]})
				}
			}
		}
		for j := i - 1; j >= 0; j-- {
			currentSum := array[i] + array[j]
			paritalSums[currentSum] = append(paritalSums[currentSum], []int{array[i], array[j]})
		}
	}

	return quadruplets
}

File: fourSum.py
from typing import List


def fourNumberSum(array:List[int],targetSum:int)->List[List[int]]:
    allPairSums  = {} # tracking the number of two sum pairs
    quadruplets = [] # return all quadrplets possible
    for i in range (1,len(array)-1):
        # try to find the first partial sum of the  target sum 
        # and index +  1 -- len(array) before computing the possible partial sum of the 
        # other aspect of the the target sum
        for j in range(i+1,len(array)):
            currentSum = array[i] +  array[j]
            difference =  targetSum - currentSum
            if difference  in allPairSums:
                for pair in allPairSums[difference]:
                    quadruplets.append(pair + [array[j],array[i]]) 
        for k in range(0,i):
            currentSum = array[k] + array[i]
            if currentSum not in allPairSums:
                allPairSums[currentSum] = [array[k],array[i]]
            else:
                allPairSums[currentSum].append([array[k],array[i]])
    return quadruplets

                
